File: TrackGenerator/TrackParts.py
# -*- coding: utf-8 -*-

"""
Contains classes of track parts such as lines, ars.
"""

import numpy as np


class TrackPart:
    Start = np.array([0, 0, 0])
    End = np.array([0, 0, 0])
    Length = -1.0


class Straight(TrackPart):
    def __init__(self, start_point, end_point):
        self.Start = start_point
        self.End = end_point
        self.Length = np.linalg.norm(end_point - start_point)


class Arc(TrackPart):
    Curvature = 0.0
    MiddlePoint = 0

    def __init__(self, start_point, end_point, curvature: float, length: float):
        self.Start = start_point
        self.End = end_point
        self.Length = length
        self.Curvature = curvature

    def __init__(self, start_point, end_point, curvature: float, length: float, middle_point=np.array([0, 0])):
        self.Start = start_point
        self.End = end_point
        self.Length = length
        self.Curvature = curvature
        self.MiddlePoint = middle_point

File: geometryCalculator/Assets/Scripts/TrackParts.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public abstract class TrackParts
{
    public Vector2 Start;
    public Vector2 End;
    public float Length;
}

public class Straight : TrackParts
{
    public Straight(Vector2 start, Vector2 end)
    {
        Start = start;
        End = end;
        Length = (end - start).magnitude;
    }
}

public class Arc : TrackParts
{
    public Arc(Vector2 start, Vector2 end, float curvature, float length)
    {
        Start = start;
        End = end;
        Curvature = curvature;
        Length = length;
    }

    public Arc(Vector2 start, Vector2 end, float curvature, float length, Vector2 middlePoint)
    {
        Start = start;
        End = end;
        Curvature = curvature;
        Length = length;
        MiddlePoint = middlePoint;
    }

    public float Curvature { get; private set; }
    public Vector2 MiddlePoint { get; private set; }
}